Development/nmos/control_protocol_ws_api.cpp

```diff
@@ -0,0 +1,537 @@
+                    {
+                        try
+                        {
+                            // extract the control protocol api version from the ws_ncp_path
+                            if (web::uri::split_path(ws_ncp_path).empty()) { throw std::invalid_argument("empty URL"); }
+                            const auto version = nmos::parse_api_version(web::uri::split_path(ws_ncp_path).back());
+
+                            // convert message to JSON
+                            const auto message = value::parse(utility::conversions::to_string_t(msg));
+
+                            // validate the base-message
+                            details::validate_controlprotocolapi_base_message_schema(version, message);
+
+                            const auto msg_type = nmos::fields::nc::message_type(message);
+                            switch (msg_type)
+                            {
+                            // See https://specs.amwa.tv/is-12/branches/v1.0.x/docs/Protocol_messaging.html#command-message-type
+                            case ncp_message_type::command:
+                            {
+                                // validate command-message
+                                details::validate_controlprotocolapi_command_message_schema(version, message);
+
+                                auto responses = value::array();
+                                auto& commands = nmos::fields::nc::commands(message);
+                                for (const auto& cmd : commands)
+                                {
+                                    const auto handle = nmos::fields::nc::handle(cmd);
+                                    const auto oid = nmos::fields::nc::oid(cmd);
+
+                                    // get methodId
+                                    const auto& method_id = nmos::details::parse_nc_method_id(nmos::fields::nc::method_id(cmd));
+
+                                    // get arguments
+                                    const auto& arguments = nmos::fields::nc::arguments(cmd);
+
+                                    value response;
+
+                                    auto resource = nmos::find_resource(resources, utility::s2us(std::to_string(oid)));
+                                    if (resources.end() != resource)
+                                    {
+                                        const auto& class_id = nmos::details::parse_nc_class_id(nmos::fields::nc::class_id(resource->data));
+
+                                        // find the relevent method handler to execute
+                                        auto method = get_control_protocol_method(class_id, method_id);
+                                        auto& nc_method_descriptor = std::get<0>(method);
+                                        auto& standard_method = std::get<1>(method);
+                                        auto& non_standard_method = std::get<2>(method);
```

Comment:
Is it worth describing the structure of the `method` variable here?
By inspection is seems that the first element is the NcMethodDescriptor, and the next two elements are the method handlers - is this structure described/explained elsewhere?

Reply:
described in control_protocol_handlers.h 

> // method definition (NcMethodDescriptor vs method handler)
> typedef std::tuple<web::json::value, standard_method_handler, non_standard_method_handler> method;

Reply:
Perhaps put exactly that wording as a a comment?
```suggestion
                                        // find the relevent method handler to execute
                                        // method tuple definition described in control_protocol_handlers.h
                                        auto method = get_control_protocol_method(class_id, method_id);
                                        auto& nc_method_descriptor = std::get<0>(method);
                                        auto& standard_method = std::get<1>(method);
                                        auto& non_standard_method = std::get<2>(method);
```